pkg/state/nodepoolhealth/tracker.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodepoolhealth
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/karpenter/pkg/utils/ringbuffer"
+)
+
+const (
+	BufferSize     = 4
+	ThresholdFalse = 0.5 // 50% of 0s for NodeRegistrationHealthy=False
+)
+
+type Status int
+
+const (
+	StatusUnknown Status = iota
+	StatusHealthy
+	StatusUnhealthy
+)
+
+type Tracker struct {
+	sync.RWMutex
+	buffer ringbuffer.RingBuffer[bool]
+}
+
+func NewTracker(capacity int) *Tracker {
+	return &Tracker{
+		buffer: *ringbuffer.New[bool](capacity),
+	}
+}
+
+func (t *Tracker) Update(success bool) {
+	t.Lock()
+	defer t.Unlock()
+
+	t.buffer.Insert(success)
+}
+
+func (t *Tracker) Reset() {
+	t.Lock()
+	defer t.Unlock()
+
+	t.buffer.Reset()
+}
+
+func (t *Tracker) Status() Status {
+	if t.buffer.Len() == 0 {
+		return StatusUnknown
+	}
+	// Count number of true and false
+	var unhealthyCount int
+	for _, value := range t.buffer.Items() {
+		if !value {
+			unhealthyCount++
+		}
+	}
+	// Determine health status based on threshold
+	if (float64(unhealthyCount) / float64(BufferSize)) >= ThresholdFalse {
+		return StatusUnhealthy
+	} else {
+		return StatusHealthy
+	}
+}
+
+func (t *Tracker) SetStatus(status Status) {
```

Comment:
This implementation only works on an empty buffer, right? We need to ensure we're always resetting the buffer first. Let's also make sure we add a unit test to catch this.